python/dgl/distributed/dist_graph.py

```diff
@@ -1229,48 +1229,38 @@ def _split_even_to_part(partition_book, elements):
-        # I hope it's OK.
-        eles = F.nonzero_1d(elements[0:len(elements)])
-        # compute the offset of each split and ensure that the difference of each partition size
-        # is 1.
-        offsets = _even_offset(len(eles), partition_book.num_partitions())
-        assert offsets[-1] == len(eles)
-
-        # Get the elements that belong to the partition.
-        partid = partition_book.partid
-        part_eles = eles[offsets[partid] : offsets[partid + 1]]
+        nonzero_count = elements.count_nonzero()
     else:
         elements = F.tensor(elements)
         nonzero_count = F.count_nonzero(elements)
-        # compute the offset of each split and ensure that the difference of each partition size
-        # is 1.
-        offsets = _even_offset(nonzero_count, partition_book.num_partitions())
-        assert offsets[-1] == nonzero_count
-
-        # Get the elements that belong to the partition.
-        partid = partition_book.partid
-        left, right = offsets[partid], offsets[partid + 1]
-
-        x = y = 0
-        num_elements = len(elements)
-        block_size = num_elements // partition_book.num_partitions()
-        part_eles = None
-        # compute the nonzero tensor of each partition instead of whole tensor to save memory
-        for idx in range(0, num_elements, block_size):
-            nonzero_block = F.nonzero_1d(elements[idx:min(idx+block_size, num_elements)])
-            x = y
-            y += len(nonzero_block)
-            if y > left and x < right:
-                start = max(x, left) - x
-                end = min(y, right) - x
-                tmp = nonzero_block[start:end] + idx
-                if part_eles is None:
-                    part_eles = tmp
-                else:
-                    part_eles = F.cat((part_eles, tmp), 0)
-            elif x >= right:
-                break
+    # compute the offset of each split and ensure that the difference of each partition size
+    # is 1.
+    offsets = _even_offset(nonzero_count, partition_book.num_partitions())
+    assert offsets[-1] == nonzero_count
+
+    # Get the elements that belong to the partition.
+    partid = partition_book.partid
+    left, right = offsets[partid], offsets[partid + 1]
+
+    x = y = 0
+    num_elements = len(elements)
+    block_size = num_elements // partition_book.num_partitions()
+    part_eles = None
+    # compute the nonzero tensor of each partition instead of whole tensor to save memory
+    for idx in range(0, num_elements, block_size):
+        nonzero_block = F.nonzero_1d(elements[idx:min(idx+block_size, num_elements)])
+        x = y
+        y += len(nonzero_block)
+        if y > left and x < right:
+            start = max(x, left) - x
+            end = min(y, right) - x
+            tmp = nonzero_block[start:end] + idx
+            if part_eles is None:
+                part_eles = tmp
+            else:
+                part_eles = F.cat((part_eles, tmp), 0)
```

Comment:
to avoid too many concatenations, can you collect all segments from different blocks first and use one concatenation in the end?

Reply:
Thanks for your suggestions, I will make a new commit for it later

Reply:
Hi, we tested on 200 machines with 10 billion edges, and the whole split process was completed in few seconds. 
If save all blocks first, and then cat, the memory usage will double in the worst case. 
So may be the current method is better?

Reply:
i see. sure, the current solution also works.